lib/srv/discovery/fetchers/db/aws.go

```diff
@@ -17,11 +17,122 @@ limitations under the License.
 package db
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
 	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/srv/discovery/common"
 )
 
+// awsFetcherPlugin defines an interface that provides database type specific
+// functions for use by the common AWS database fetcher.
+type awsFetcherPlugin interface {
+	// GetDatabases fetches databases from AWS API and converts the results to
+	// Teleport types.Databases.
+	GetDatabases(context.Context, *awsFetcherConfig) (types.Databases, error)
+	// ComponentShortName provides the plugin's short component name for
+	// logging purposes.
+	ComponentShortName() string
+}
+
+// awsFetcherConfig is the AWS database fetcher configuration.
+type awsFetcherConfig struct {
+	// AWSClients are the AWS API clients.
+	AWSClients cloud.AWSClients
+	// Type is the type of DB matcher, for example "rds", "redshift", etc.
+	Type string
+	// AssumeRole provides a role ARN and ExternalID to assume an AWS role
+	// when fetching databases.
+	AssumeRole types.AssumeRole
+	// Labels is a selector to match cloud database tags.
+	Labels types.Labels
+	// Region is the AWS region selector to match cloud databases.
+	Region string
+	// Log is a field logger to provide structured logging for each matcher,
+	// based on its config settings by default.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates the config and sets defaults.
+func (cfg *awsFetcherConfig) CheckAndSetDefaults(component string) error {
+	if cfg.AWSClients == nil {
+		return trace.BadParameter("missing parameter AWSClients")
+	}
+	if cfg.Type == "" {
+		return trace.BadParameter("missing parameter Type")
+	}
+	if len(cfg.Labels) == 0 {
+		return trace.BadParameter("missing parameter Labels")
+	}
+	if cfg.Region == "" {
+		return trace.BadParameter("missing parameter Region")
+	}
+	if cfg.Log == nil {
+		cfg.Log = logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:" + component,
+			"labels":        cfg.Labels,
+			"region":        cfg.Region,
+			"role":          cfg.AssumeRole,
+		})
+	}
+	return nil
+}
+
+// newAWSFetcher returns a AWS database fetcher for the provided selectors
+// and AWS database-type specific fetcher plugin.
+func newAWSFetcher(cfg awsFetcherConfig, plugin awsFetcherPlugin) (*awsFetcher, error) {
+	if err := cfg.CheckAndSetDefaults(plugin.ComponentShortName()); err != nil {
```

Comment:
Can we just used the cfg.Type : 
> // Type is the type of DB matcher, for example "rds", "redshift", etc.

and get ride of `ComponentShortName()` ?

Reply:
this is to preserve the prior component name used in logs. We had shorter names than the matcher type in logs, because the logger truncates them, e.g. rather than "redshift-serverless" we used "rss<".

Reply:
without doing this, there is no logger component name difference between, e.g. "redshift" and "redshift-serverless" due to truncation.